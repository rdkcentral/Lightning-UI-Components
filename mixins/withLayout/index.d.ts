import lng from '@lightningjs/core';

export interface LayoutOptions {
  w?: number;
  h?: number;
  circle?: boolean;
  ratioX?: number;
  ratioY?: number;
  upCount?: number;
}

export interface WithLayout {
  _previousDimensionsData: string;
  _itemLayout?: LayoutOptions;

  set itemLayout(val: LayoutOptions);
  get itemLayout(): LayoutOptions;

  _allowUpdate(): boolean;
  _updateItemLayout(): void;
}

export type WithLayoutConstructor<BaseType extends typeof lng.Component> =
  BaseType & {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    new (...args: any[]): BaseType & WithLayout;
  };

export default function withLayout<BaseType extends typeof lng.Component>(
  base: BaseType
): WithLayoutConstructor<BaseType>;
