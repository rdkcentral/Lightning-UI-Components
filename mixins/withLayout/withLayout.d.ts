export interface LayoutOptions {
  w?: number;
  h?: number;
  circle?: boolean;
  ratioX?: number;
  ratioY?: number;
  upCount?: number;
}

export interface WithLayout {
  _previousDimensionsData: string;
  _itemLayout?: LayoutOptions;

  set itemLayout(val: LayoutOptions): void;
  get itemLayout(): LayoutOptions;

  _allowUpdate(): boolean;
  _updateItemLayout(): void;
}

export interface WithLayout {
  new(...args: any[]): WithLayout;
}

export default function withLayout<T>(
  base: T,
): T & WithAnnouncerConstructor;