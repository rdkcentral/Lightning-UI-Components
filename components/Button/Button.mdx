import { Preview, Story } from '@storybook/addon-docs/blocks';
// uncomment the line below and edit the path to point to your component
// import * as stories from './Button.stories';

# Button

Basic component that handles styling and functionality for a button

## Usage

```js
import { Button } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      Button: {
        type: Button,
        title: 'Button'
      }
    };
  }
};
```

Before the button receives its title, it should be in a loading state.

<Preview>
  <Story id="button--basic" />
</Preview>


### Sizing

Buttons can take in a`w`. This is the minimum width that a button will be, the button will dynamically size itself based on the width of the title provided. Buttons can also take in a `h` - if not provided it will default to `40px`.

### Focus

When focused, the background/foreground colors are inverted by default. This can be overridden with theming.

### Icon

All buttons have the ability to be passed an icon. This will position the icon to the left of the title and center it horizontally and vertically.

### Theme

Themes have the ability to override default style properties for button mount and focus and unfocus states.
Focus and unfocus can take in patch functions that update styles on button _focus and _unfocus. If a theme property is also set directly on the Button component, the property set on the component itself will override the theme.

```js
  radius: 0,
  text: { fontSize: 20 },
  w: 150,
  h: 40,
  padding: 50,
  unfocus: {
    background: 0xff1f1f1f,
    text: 0xffffffff,
    icon: 0xffffffff,
    patch: function() {
      this.patch({
        smooth: { color: this._theme.unfocus.background },
        Content: {
          Title: { smooth: { color: this._theme.unfocus.text } },
          Icon: { smooth: { color: this._theme.unfocus.icon } }
        }
      });
    }
  },
  focus: {
    background: 0xffffffff,
    radius: 0,
    text: 0xff1f1f1f,
    icon: 0xff1f1f1f,
    patch: function() {
      this.patch({
        smooth: { color: this._theme.focus.background },
        Content: {
          Title: { smooth: { color: 0xff1f1f1f } },
          Icon: { smooth: { color: this._theme.focus.icon } }
        }
      });
    }
  }
}
```




## API

### Properties

name|type|readonly|description
--|--|--|--
announce|string|true|announcement text for screen readers
h|number|false|height of the button
w|number|false|width of the button
title|string|false|button text
radius|rumber|false|corner radius for a button
stroke|object|false|`color` and `weight` properties for button stroke
loading|object|false|`tag` and `animation` properties for button loading animation
icon|object|false|`src`, `size`, and `spacing` properties for an icon
theme|object|false|properties that apply style to the button

### Methods

#### _handleEnter

Can be overridden to apply handling for the <kbd>Enter</kbd> key
