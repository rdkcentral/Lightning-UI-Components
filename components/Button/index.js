/**
 * Button Component
 *
 * Generic Button Component that handles setting text
 * and focus and unfocus states for the button background.
 */
import lng from 'wpe-lightning';
import { RoundRect } from '../../utils';
import withStyles from '../../mixins/withStyles';
import Icon from '../Icon';

export const styles = {
  w: 150,
  h: 40,
  radius: 0,
  background: { color: 0xff1f1f1f },
  icon: { color: 0xffffffff },
  text: {
    fontSize: 20,
    color: 0xffffffff
  },
  padding: 50,
  unfocus: {
    patch: function() {
      const { background, icon, text } = this.styles;
      this.patch({
        smooth: { color: background.color },
        Content: {
          Title: { smooth: { color: text.color } },
          Icon: { smooth: { color: icon.color } }
        }
      });
    }
  },
  focus: {
    background: 0xffffffff,
    text: 0xff1f1f1f,
    icon: 0xff1f1f1f,
    patch: function() {
      const { background, icon, text } = this.styles.focus;
      this.patch({
        smooth: { color: background },
        Content: {
          Title: { smooth: { color: text } },
          Icon: { smooth: { color: icon } }
        }
      });
    }
  }
};

class Button extends lng.Component {
  static _template() {
    return {
      w: this.styles.w,
      h: this.styles.h,
      radius: this.styles.radius,
      Content: {
        mount: 0.5,
        x: w => w / 2,
        y: h => h / 2,
        flex: {
          direction: 'row',
          alignContent: 'center',
          alignItems: 'center'
        },
        Icon: {
          type: Icon
        },
        // TODO: Wonky lineHeight / fontsize from Lightning
        // Move title down 2 pixels to _visually_ center it
        // inside the button
        Title: { y: 2 }
      },
      Stroke: {
        zIndex: -1,
        mount: 0.5,
        x: w => w / 2,
        y: h => h / 2
      }
    };
  }

  _init() {
    this._update();
  }

  _focus() {
    this.styles.focus.patch.apply(this);
  }

  _unfocus() {
    this.styles.unfocus.patch.apply(this);
  }

  _update() {
    const template = { Content: {} };
    const Title = {};
    const Icon = {};
    const Stroke = {};

    const radius = this.radius || this.styles.radius;

    template.color = [this.background, this.styles.background.color].find(
      Number.isFinite
    );

    if (this.title) {
      Title.text = {
        ...this.styles.text,
        fontColor: this.styles.text.color,
        fontSize: this.fontSize || this.styles.text.fontSize,
        text: this.title
      };
      Title.color = this.styles.text.color;
      template.Content.Title = Title;
    }

    if (this.icon) {
      const { color, size, spacing, src } = this.icon;
      Icon.color = color || this.styles.icon.color;
      Icon.w = size;
      Icon.h = size;
      Icon.flexItem = { marginRight: spacing };
      Icon.icon = src;
      template.Content.Icon = Icon;
    }

    if (this.stroke) {
      const { color, weight } = this.stroke;
      const radius = this.radius || this.styles.radius;

      template.texture = lng.Tools.getRoundRect(
        RoundRect.getWidth(this.w),
        RoundRect.getHeight(this.h),
        radius,
        0x00,
        true,
        0xffffffff
      );

      Stroke.color = color;
      Stroke.texture = lng.Tools.getRoundRect(
        RoundRect.getWidth(this.w),
        RoundRect.getHeight(this.h),
        radius,
        weight,
        0xffffffff,
        true,
        this.background
      );

      template.Stroke = Stroke;
    } else {
      template.texture = lng.Tools.getRoundRect(
        RoundRect.getWidth(this.w),
        RoundRect.getHeight(this.h),
        radius
      );
    }

    this.patch(template);

    this._Title.on('txLoaded', () => {
      let iconSize = this._icon ? this._icon.size + this._icon.spacing : 0;
      let padding = [this.padding, this.styles.padding, 50].find(
        Number.isFinite
      );
      if (!this.fixed) {
        const w = this._Title.renderWidth + padding * 2 + iconSize;
        if (w !== this.w) {
          this.w = w > this.w ? w : this.w;
          this.fireAncestors('$itemChanged');
          this.signal('buttonWidthChanged', { w: this.w });
          this._update();
        }
      }
    });
  }

  _handleEnter() {
    if (typeof this.onEnter === 'function') {
      this.onEnter(this);
    }
  }

  get title() {
    return this._title;
  }

  set title(title) {
    this._title = title;
    this._update();
  }

  get icon() {
    return this._icon;
  }

  set icon({ src, size = 20, spacing = 5, color = 0xffffffff }) {
    if (src) {
      this._icon = { src, size, spacing, color };
      this._update();
    }
  }

  get stroke() {
    return this._stroke || this.styles.stroke;
  }

  set stroke({ weight = 2, color = 0x00 }) {
    this._stroke = { weight, color };
    this._update();
  }

  get announce() {
    // TODO - Localization?
    // Do we need a locale file with
    // component translations?
    return this._title + ', Button';
  }

  get _Title() {
    return this.tag('Content.Title');
  }
  get _Icon() {
    return this.tag('Content.Icon');
  }
  get _Stroke() {
    return this.tag('Stroke');
  }
}

export default withStyles(Button, styles);
