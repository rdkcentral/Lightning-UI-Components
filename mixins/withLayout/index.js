import { default as context } from '../../context';
import { getDimensions } from '../../Styles/Layout';
import { stringifyCompare } from '../../utils';

export default function withLayout(Base) {
  return class extends Base {
    _construct() {
      super._construct && super._construct();
      this._previousDimensionData = null; // Store the previous dimension data to compare next time _updateItemLayout is called
    }

    get itemLayout() {
      return this._itemLayout;
    }

    set itemLayout(itemLayout) {
      if (!stringifyCompare(this._itemLayout, itemLayout)) {
        this._itemLayout = itemLayout;
        this._updateItemLayout();
      }
    }

    _allowUpdate() {
      const {
        w = '',
        h = '',
        circle = '',
        ratioX = '',
        ratioY = '',
        upCount = ''
      } = this._itemLayout || {};
      // Store a string representation of all values from previous update to compare and decide if layout should change
      const layoutString =
        Object.values(context.theme.layout).join('') +
        `${w}${h}${circle ? 1 : 0}${ratioX}${ratioY}${upCount}`;
      if (layoutString !== this._previousDimensionData) {
        this._previousDimensionData = layoutString;
        return true;
      }
      return false;
    }

    _updateItemLayout() {
      if (!this._allowUpdate()) return;
      const { w, h } = getDimensions(this._itemLayout);
      if (this._itemLayout && this._itemLayout.circle) {
        this.w = this.h;
        this._layoutCircleMode = true;
        if (this.style) this.style.radius = this.h / 2;
      } else if (h || w) {
        // If there is not enough information passed in args to calculate item size
        const width = context.theme.layout.screenW;
        const height = context.theme.layout.screenH;
        this.h = h || w * (height / width);
        this.w = w || h * (width / height);
      }
      if (
        !this._itemLayout ||
        (!this._itemLayout.circle && this._layoutCircleMode)
      ) {
        // Cleanup after circle layout removed
        this.style.radius = undefined;
        this._layoutCircleMode = false;
      }
      super._update && super._update();
      this.fireAncestors('$itemChanged');
    }
  };
}
