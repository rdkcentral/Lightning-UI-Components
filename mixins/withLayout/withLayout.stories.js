import lng from '@lightningjs/core';
import withLayout from '.';
import mdx from './withLayout.mdx';
import parks from '../../assets/images/Parks_and_Recreation_16x9.jpg';

import withMetadata from '../../patterns/withMetadata';
import { Column } from '../../layout';
import { MetadataTile } from '../../elements';

export default {
  title: 'Mixins/withLayout',
  parameters: {
    docs: {
      page: mdx
    },
    stage: { w: 800, h: 600 }
  }
};

export const Basic = args => {
  return class Basic extends lng.Component {
    static _template() {
      return {
        Wrapper: {
          type: lng.Component,
          h: 600,
          w: 800,
          Column: {
            h: 600,
            type: Column,
            itemSpacing: 100,
            items: [
              {
                type: withLayout(withMetadata()),
                src: parks,
                w: 320,
                h: 180,
                itemLayout: {
                  w: args.w,
                  h: args.h
                },
                Metadata: {
                  type: MetadataTile,
                  firstLine: `${args.h || 'default'} height, ${
                    args.w || 'default'
                  } width`
                }
              },
              {
                type: withLayout(withMetadata()),
                src: parks,
                w: 320,
                h: 180,
                itemLayout: {
                  h: args.h,
                  ratioX: args.ratioX,
                  ratioY: args.ratioY
                },
                Metadata: {
                  type: MetadataTile,
                  firstLine: ` ${args.h || 'default'} height, ${args.ratioX}X${
                    args.ratioY
                  }`
                }
              },
              {
                type: withLayout(withMetadata()),
                src: parks,
                w: 320,
                h: 180,
                itemLayout: {
                  ratioX: args.ratioX,
                  ratioY: args.ratioY,
                  upCount: args.upCount
                },
                Metadata: {
                  type: MetadataTile,
                  firstLine: `${args.ratioX}X${args.ratioY}, ${args.upCount} upCount`
                }
              },
              {
                type: withLayout(withMetadata()),
                src: parks,
                w: 320,
                h: 180,
                itemLayout: {
                  h: args.h,
                  upCount: args.upCount
                },
                Metadata: {
                  type: MetadataTile,
                  firstLine: `${args.h || 'default'} height, ${
                    args.upCount
                  } upCount`
                }
              },
              {
                type: withLayout(withMetadata()),
                src: parks,
                w: 320,
                h: 180,
                itemLayout: {
                  h: args.h
                },
                Metadata: {
                  type: MetadataTile,
                  firstLine: `${args.h || 'default'} height only`
                }
              },
              {
                type: withLayout(withMetadata()),
                src: parks,
                w: 320,
                h: 180,
                itemLayout: {
                  w: args.w
                },
                Metadata: {
                  type: MetadataTile,
                  firstLine: `${args.w || 'default'} width only`
                }
              }
            ]
          }
        }
      };
    }

    _getFocused() {
      return this.tag('Wrapper.Column');
    }
  };
};

Basic.args = {
  w: 340,
  h: 180,
  ratioX: 16,
  ratioY: 9,
  upCount: 4
};

Basic.argTypes = {
  w: { control: 'number' },
  h: { control: 'number' },
  ratioX: { control: 'number' },
  ratioY: { control: 'number' },
  upCount: { control: 'number' }
};
