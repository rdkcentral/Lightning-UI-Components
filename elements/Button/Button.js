/**
 * Button Component
 *
 * Generic Button Component that handles setting text icon and checkbox
 */
import { withExtensions } from '../../mixins';
import Surface from '../Surface';
import TextBox from '../TextBox';
import * as styles from './Button.styles';
import { measureTextWidth } from '../../utils';
import Row from '../../layout/Row';

class Button extends Surface {
  static get __componentName() {
    return 'Button';
  }

  static get __themeStyles() {
    return styles;
  }

  static get properties() {
    return ['fixed', 'justify', 'prefix', 'suffix', 'title'];
  }

  static get tags() {
    return [
      ...super.tags,
      'Content',
      { name: 'Title', path: 'Content.Title' },
      { name: 'Prefix', path: 'Content.Prefix' },
      { name: 'Suffix', path: 'Content.Suffix' }
    ];
  }

  static _template() {
    return {
      ...super._template(),
      Content: {
        mount: 0.5,
        x: w => w / 2,
        y: h => h / 2
      }
    };
  }

  _update() {
    this._updatePrefix();
    this._updateTitle();
    this._updateSuffix();
    this._updateAllPositioning();
    this._updateTruncation();
  }

  _updateAllPositioning() {
    this._updatePositions();
    this._updateContentDimensions();
    this._updateSurfaceDimensions();
    this._updateContentPosition();
    // delaying the super._update call so we update the surface sizing after the button has finished updating
    super._update();
  }

  $itemChanged() {
    this._updateAllPositioning();
  }

  _onTextBoxChanged() {
    this._updateAllPositioning();
  }

  _updatePositions() {
    if (this._hasPrefix) {
      this._Prefix.x = this._prefixX;
    }
    if (this._hasTitle) {
      this._Title.x = this._titleX;
    }
    if (this._hasSuffix) {
      this._Suffix.x = this._suffixX;
    }
  }

  _updatePrefix() {
    if (this.prefix) {
      let prefixPatch = {
        h: this.style.prefixH,
        itemSpacing: this.style.itemSpacing
      };
      if (!this._Prefix) {
        prefixPatch = {
          ...this._rowProps,
          ...prefixPatch
        };
      }
      this._Content.patch({ Prefix: prefixPatch });
      this._Prefix.items = this._addButtonProps(this.prefix);
    } else {
      this._Content.patch({ Prefix: undefined });
    }
  }

  _updateTitle() {
    if (this._hasTitle) {
      let titlePatch = {
        content: this.title,
        textStyle: {
          textColor: this.style.contentColor,
          ...this.style.textStyle
        }
      };
      if (!this._Title) {
        titlePatch = {
          type: TextBox,
          mountY: 0.5,
          y: h => h / 2,
          signals: {
            textBoxChanged: '_onTextBoxChanged'
          },
          ...titlePatch
        };
      }
      this._Content.patch({ Title: titlePatch });
    } else {
      this.patch({ Content: { Title: undefined } });
    }
  }

  _updateSuffix() {
    if (this.suffix) {
      let suffixPatch = {
        h: this.style.suffixH,
        itemSpacing: this.style.itemSpacing
      };
      if (!this._Suffix) {
        suffixPatch = {
          ...this._rowProps,
          ...suffixPatch
        };
      }
      this._Content.patch({ Suffix: suffixPatch });
      this._Suffix.items = this._addButtonProps(this.suffix);
    } else {
      this._Content.patch({ Suffix: undefined });
    }
  }

  _updateTruncation() {
    /**
     * needs to run before rest of content updates to prevent a bug causing text
     * to sometimes truncate and then un-truncate(with no width change) when
     * toggling fixed after toggling focus
     */
    if (this._Title) {
      this._Title.patch({
        wordWrap: this.fixed,
        wordWrapWidth: this.fixed ? this.w - this._paddingX : 0
      });
    }
  }

  _updateContentDimensions() {
    const contentDimensionsPatch = {};
    const y = this.h / 2;

    // reference transition targets to ensure we aren't checking against intermittent values
    if (this._Content.transition('w').targetValue !== this._contentW) {
      contentDimensionsPatch.w = this._contentW;
    }

    if (this._Content.y !== y) {
      contentDimensionsPatch.y = y;
    }

    if (Object.keys(contentDimensionsPatch).length > 0) {
      if (this._smooth) {
        this._Content.smooth = contentDimensionsPatch;
      } else {
        this._Content.patch(contentDimensionsPatch);
      }
    }
  }

  _updateContentPosition() {
    let mountX;
    let x;

    switch (this.justify) {
      case 'left':
        mountX = 0;
        x = this._paddingLeft;
        break;
      case 'right':
        mountX = 1;
        x = this.w - this._paddingRight;
        break;
      case 'center':
      default:
        mountX = 0.5;
        x = this.w / 2;
        break;
    }

    this._Content.patch({ mountX, x });
  }

  _updateSurfaceDimensions() {
    let newWidth = this.w;
    if (this.fixed) {
      newWidth = this._w;
    } else {
      // if no title, ignore minWidth and use prefix/suffix width
      newWidth =
        !this._hasTitle && (this._hasPrefix || this._hasSuffix)
          ? this._contentW + this._paddingX
          : Math.max(this._contentW + this._paddingX, this.style.minWidth);
    }

    if (newWidth !== this.w) {
      this.w = newWidth;
      this.fireAncestors('$itemChanged');
    }
  }

  _addButtonProps(arr) {
    const items = Array.isArray(arr) ? arr : [arr];
    return items.map(item => ({
      centerInParent: true,
      palette: this.palette,
      mode: this.mode,
      ...item,
      style: {
        color: this.style.contentColor,
        ...item.style
      }
    }));
  }

  _getJustify() {
    return !!this._justify ? this._justify : this.style.justify;
  }

  get _hasPrefix() {
    return this._Prefix && Array.isArray(this.prefix)
      ? this.prefix.length
      : this.prefix;
  }

  get _prefixW() {
    return this._hasPrefix ? this._Prefix.w : 0;
  }

  get _prefixX() {
    return 0;
  }

  get _hasTitle() {
    return !!this.title;
  }

  get _titleW() {
    // get the pre-calculated text width of the title to solve some race condition bugs
    if (
      this._hasTitle &&
      this._Title &&
      this._Title._Text &&
      this._Title.visible
    ) {
      if (this.fixed) {
        return this._Title.w;
      } else {
        return measureTextWidth(this._Title._Text.text);
      }
    }
    return 0;
  }

  get _titleX() {
    return this._hasPrefix ? this._prefixW + this.style.titlePadding : 0;
  }

  get _hasSuffix() {
    return this._Suffix && Array.isArray(this.suffix)
      ? this.suffix.length
      : this.suffix;
  }

  get _suffixW() {
    return this._hasSuffix ? this._Suffix.w : 0;
  }

  get _suffixX() {
    if (this._hasTitle) {
      return this._titleW + this._Title.x + this.style.titlePadding;
    } else if (this._hasPrefix) {
      return this._prefixW + this.style.itemSpacing;
    }
    return 0;
  }

  get _contentW() {
    if (this._hasSuffix) {
      return this._suffixX + this._suffixW;
    } else if (this._hasTitle && this._Title.visible) {
      return this._titleX + this._titleW;
    } else if (this._hasPrefix) {
      return this._prefixX + this._prefixW;
    }
    return 0;
  }

  get _rowProps() {
    return {
      type: Row,
      mountY: 0.5,
      autoResizeWidth: true
    };
  }

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    if (this._announce) {
      return this._announce;
    }

    // TODO - Localization?
    // Do we need a locale file with
    // component translations?
    return this.title + ', Button';
  }

  get _paddingX() {
    return this._paddingLeft + this._paddingRight;
  }

  get _paddingLeft() {
    return this._hasTitle ? this.style.paddingX : this.style.paddingXNoTitle;
  }

  get _paddingRight() {
    return this._hasTitle ? this.style.paddingX : this.style.paddingXNoTitle;
  }
}

export default withExtensions(Button);
