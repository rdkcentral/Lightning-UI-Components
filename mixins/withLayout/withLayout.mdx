<!--
 Copyright 2021 Comcast Cable Communications Management, LLC

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

 SPDX-License-Identifier: Apache-2.0
 -->
 
import { Canvas, Story } from '@storybook/addon-docs/blocks';

# withLayout

Use `withLayout` to autotomatically size an item according to screen dimensions, upCount and/or set explicitly.

## Usage

Let's use an example to see how we can use `withLayout`. Let's say I have an a `Tile` component on screen that should be sized according the screen size:

```js
import lng from '@lightningjs/core';
import { withLayout } from '@lightningjs/mixins';

class Basic extends lng.Component {
  static _template() {
    return {
      Tile: {
        type: withLayout(Tile),
        src: parks,
        itemLayout: {
          ratioX: 16,
          ratioY: 9,
          upCount: 5
        }
      }
    };
  }
}
```

If I want to explicitly set a height but the only other information is the screen dimensions, I can do that by adding these parameters to the itemLayout object.

```js
import lng from '@lightningjs/core';
import { withLayout } from '@lightningjs/mixins';

class Basic extends lng.Component {
  static _template() {
    return {
      Tile: {
        type: withLayout(Tile),
        src: parks,
        itemLayout: {
          ratioX: 16,
          ratioY: 9,
          h: 180
        }
      }
    };
  }
}
```

If I am unsure of the width, I can autosize by adding a height and upCount properties to the itemLayout object.

```js
import lng from '@lightningjs/core';
import { withLayout } from '@lightningjs/mixins';

class Basic extends lng.Component {
  static _template() {
    return {
      Tile: {
        type: withLayout(Tile),
        src: parks,
        itemLayout: {
          h: 320,
          upCount: 4,
        }
      }
    };
  }
}
```

**NOTE: One of the following combinations of properties is required:**

- `h` and `w` (If explicitly set, it may be uncessary to use the `withLayout` pattern as these properties can be set on the component),
- `h` and `ratioX` and `ratioY`,
- `ratioX` and `ratioY` and `upCount`,
- `h` and `upCount`,
- `h`
- `w`

Auto sizing of the item will happen based on the above priority. So if the `w` and `h` are passed no other calculations will occur.
If the mixin is unable to calculate the height OR width from the passed parameters. The height will default 9/16 of the width. Inversely the width will default to 16/9 of the height.
If the mixin is unable to calculate height AND width from the passed parameters. No autosize will occur.

<Canvas>
  <Story id="mixins-withlayout--basic" />
</Canvas>

### Properties

name|type|required|default|description
--|--|--|--|--
itemLayout|object|true|-|the auto sizing properties that will override explicitly set h/w for an item

### itemLayout

name|type|required|default|description
--|--|--|--|--
w|number|false|-|explicitly set the width of an item
h|number|false|-|explicitly set the height of an item
ratioX|number|false|-|units of x resolution relative to the screen height
ratioY|number|false|-|units of y resolution relative to the screen width
upCount|number|false|-|the number of items that should be displayed within screen bounds. Additional items will run off-screen or outside of specified row width
