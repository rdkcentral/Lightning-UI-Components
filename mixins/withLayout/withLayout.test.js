/**
 * Copyright 2021 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import TestUtils from '../../test/lightning-test-utils';
import Button from '../../elements/Button';
import { SCREEN } from '../../Styles/Layout';
import withLayout from '.';

const createButtonWithLayout = props =>
  TestUtils.makeCreateComponent(withLayout(Button))(props);

describe('withLayout', () => {
  let buttonWithLayout, testRenderer;
  const w = 320;
  const h = 180;
  const ratioY = 9;
  const ratioX = 16;
  const upCount = 5;

  beforeEach(() => {
    [buttonWithLayout, testRenderer] = createButtonWithLayout();
  });
  afterEach(() => {
    buttonWithLayout = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('should explicity set height, width', () => {
    buttonWithLayout.itemLayout = { h, w };
    testRenderer.update();
    expect(buttonWithLayout.h).toBe(h);
    expect(buttonWithLayout.w).toBe(w);
    expect(buttonWithLayout.itemLayout).toStrictEqual({ h: 180, w: 320 });
  });

  it('should set width from height and screen dimension', () => {
    buttonWithLayout.itemLayout = { h, ratioX, ratioY };
    testRenderer.update();
    expect(buttonWithLayout.h).toBe(h);
    expect(buttonWithLayout.w).toBe((h * 16) / 9);
  });

  it('should set width from height and upCount', () => {
    buttonWithLayout.itemLayout = { h, upCount };
    testRenderer.update();
    expect(buttonWithLayout.w).toBeDefined();
    expect(buttonWithLayout.w).not.toBe(0);
  });

  it('should set height and width from screen dimensions and upCount', () => {
    buttonWithLayout.itemLayout = { ratioX, ratioY, upCount };
    testRenderer.update();
    expect(buttonWithLayout.w).toBeDefined();
    expect(buttonWithLayout.w).not.toBe(0);
  });

  it('should set height from width and default ratio', () => {
    buttonWithLayout.itemLayout = { w };
    testRenderer.update();
    expect(buttonWithLayout.h).toBeDefined();
    expect(buttonWithLayout.h).toBe((w * SCREEN.h) / SCREEN.w);
  });

  it('should set width from height and default ratio', async () => {
    buttonWithLayout.itemLayout = { h };
    testRenderer.update();
    expect(buttonWithLayout.w).toBeDefined();
    expect(buttonWithLayout.w).toBe((h * SCREEN.w) / SCREEN.h);
  });
});
