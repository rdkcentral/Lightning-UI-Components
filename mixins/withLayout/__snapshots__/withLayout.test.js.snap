// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withLayout renders 1`] = `
Object {
  "Component": Object {
    "Artwork": Object {
      "active": false,
      "alpha": 0.001,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Image": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Image",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Image",
          ],
          "texture": Object {
            "src": "./images/default_background.png",
            "type": "ImageTexture",
          },
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 1,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "hasFinalFocus": false,
      "hasFocus": false,
      "isComponent": true,
      "mount": 0.5,
      "mountX": 0.5,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Artwork",
      "renderOfScreen": undefined,
      "renderToTexture": true,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "tags": Array [
        "Artwork",
      ],
      "type": "Artwork",
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
    "Background": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Background",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Background",
      ],
      "texture": Object {
        "type": "StaticCanvasTexture",
      },
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
